mobile-browser-based-version/server/src/logic/federated/handlers.js

```diff
@@ -0,0 +1,547 @@
+import path from 'path';
+import fs from 'fs';
+import msgpack from 'msgpack-lite';
+import * as config from '../../../server.config.js';
+import {
+  averageWeights,
+  assignWeightsToModel,
+} from '../../helpers/tfjs_helpers.js';
+import tasks from '../../tasks/tasks.js';
+import * as tf from '@tensorflow/tfjs';
+import '@tensorflow/tfjs-node';
+
+const REQUEST_TYPES = Object.freeze({
+  CONNECT: 'connect',
+  DISCONNECT: 'disconnect',
+  SELECTION_STATUS: 'selection-status',
+  AGGREGATION_STATUS: 'aggregation-status',
+  POST_WEIGHTS: 'post-weights',
+  GET_WEIGHTS: 'get-weights',
+  POST_SAMPLES: 'post-samples',
+  GET_SAMPLES: 'get-samples',
+  GET_TASKS: 'get-tasks',
+});
+/**
+ * Fraction of client gradients required on final round of federated training
+ * to proceed to the pooling step.
+ */
+const AGGREGATION_THRESHOLD = 0.8;
+/**
+ * Number of selected clients required to start federated training.
+ */
+const TRAINING_THRESHOLD = 2;
+/**
+ * Once a certain threshold has been hit, leave a small window for late clients
+ * to join the next federated training session.
+ */
+const TRAINING_COUNTDOWN = 1000 * 10;
+const AGGREGATION_COUNTDOWN = 1000 * 3;
+/**
+ * Contains the model weights received from clients for a given task and round.
+ * Stored by task ID, round number and client ID.
+ */
+const weightsMap = new Map();
+/**
+ * Contains the number of data samples used for training by clients for a given
+ * task and round. Stored by task ID, round number and client ID.
+ */
+const dataSamplesMap = new Map();
```

Comment:
the name 'samples' is a bit confusing here. couldn't we make a map containing all `metadata` which clients send to server in the round (including model version which they received from server and computed gradients on (so that's the previous round number), number of samples and local steps done since then, and possibly other stuff so it's extensible?)